docs/tools/testing/unit_testing.md

```diff
@@ -1,27 +1,275 @@
+	rtos/Semaphore/test_Semaphore.cpp
+)
+```
+3. Create a test source file `UNITTESTS/rtos/Semaphore/test_Semaphore.cpp` with the following content:
+```
+#include  "gtest/gtest.h"
+#include  "rtos/Semaphore.h"
+
+static osStatus_t retval = osOK;
+static uint32_t count = 0;
+
+// Test stubs
+osStatus_t osSemaphoreAcquire(osSemaphoreId_t semaphore_id, uint32_t timeout)
+{
+    return retval;
+}
+osStatus_t osSemaphoreDelete(osSemaphoreId_t semaphore_id)
+{
+    return retval;
+}
+osStatus_t osSemaphoreRelease(osSemaphoreId_t semaphore_id)
+{
+    return retval;
+}
+uint32_t osSemaphoreGetCount(osSemaphoreId_t semaphore_id)
+{
+    return count;
+}
+osSemaphoreId_t osSemaphoreNew(uint32_t max_count, uint32_t initial_count, const osSemaphoreAttr_t *attr)
+{
+    return (void *)&count; // Just a dymmy reference
+}
+
+class  TestSemaphore : public  testing::Test {
+protected:
+    rtos::Semaphore *sem;
+
+    virtual  void  SetUp()
+    {
+        sem = new rtos::Semaphore();
+    }
+
+    virtual  void  TearDown()
+    {
+        delete sem;
+    }
+};
+
+TEST_F(TestSemaphore, constructor)
+{
+    EXPECT_TRUE(sem);
+}
+```
+
+### Building and running unit tests
+
+#### Testing with Mbed CLI
+
+You can build and run unit tests through Arm Mbed CLI. Mbed CLI supports unit testing through `mbed test --unittests` command with the same arguments. You can also use the tool to generate new test files. To learn how to do so, please see [a handbook section about using unit tests with Mbed CLI](https://os.mbed.com/docs/latest/tools/test-and-debug.html#unit-testing). For other information on using Mbed CLI, please see the [CLI documentation](https://os.mbed.com/docs/latest/tools/arm-mbed-cli.html).
+
+#### Build manually without Python tools
+
+1. Create a build directory e.g. `mkdir UNITTESTS/build`.
+2. Move to the build directory `cd UNITTESTS/build`.
+3. Run CMake with `cmake [RELATIVE PATH TO UNITTESTS DIR] [OPTIONAL ARGUMENTS]` e.g. `cmake ..`:
+	* Add `-g [generator]` argument if target other than Unix Makefiles e.g. MinGW `-g "MinGW Makefiles"`
+4. Run make program (make, gmake, mingw32-make, etc).
+
+##### Run CMake and build Unix Makefiles (GNU make)
+
+```
+cmake ..
+make
+```
+
+##### Run CMake and build MinGW Makefiles (mingw32-make)
+
+```
+cmake -G "MinGW Makefiles" ..
+mingw32-make
+```
+
+##### Custom CMake variables
+
+Usage: 
+`cmake [RELATIVE PATH TO UNITTESTS DIR] [OPTIONS]`
+
+Keyword variables (usage `cmake -D<VARIABLE>(:<TYPE>)=<value>`:
+
+| Variable | Type | Accepted values | Description |
+| -------- | ---- | --------------- | ----------- |
+| COVERAGE | STRING | merged<br>separate | Generate merged or individual reports |
+
+#### Run in terminal
+
+Unit tests can be run separately from each executable or by using ctest test runner. Run ctest with make program using target test. Options can be passed to ctest using ARGS argument. See [ctest manual](https://cmake.org/cmake/help/v3.0/manual/ctest.1.html) for more information.
```

Comment:
I don't understand the "or" in the first sentence. I think it's because there is nothing tying the title and the two options in that sentence together.

Reply:
Why do you keep calling it "make program" when ctest isn't specified as a program? What does the word "program" contribute?

Reply:
CMake and CTest; please correct everywhere except commands